File: src/cspfj/constraint/extension/ExtensionConstraint2D.java
/**
 * CSPFJ - CSP solving API for Java
 * Copyright (C) 2006 Julien VION
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

package cspfj.constraint.extension;

import java.util.Collection;
import java.util.List;

import cspfj.constraint.AbstractPVRConstraint;
import cspfj.constraint.DynamicConstraint;
import cspfj.problem.Variable;

public class ExtensionConstraint2D extends AbstractPVRConstraint implements
		ExtensionConstraint, DynamicConstraint {

	// private final double tightness;

	protected MatrixManager2D matrix;

	public ExtensionConstraint2D(final Variable[] scope, final Matrix2D matrix,
			final boolean shared) {
		this(scope, matrix, shared, false);
	}

	public ExtensionConstraint2D(Variable[] scope, Matrix2D matrix,
			boolean shared, boolean emptyMatrix) {

		super(scope);
		this.matrix = new MatrixManager2D(scope, matrix, shared);
		this.matrix.setTuple(tuple);

		if (emptyMatrix) {
			nbInitConflicts = new long[getArity()][];
			nbMaxConflicts = new long[getArity()];
			for (int i = getArity(); --i >= 0;) {
				nbInitConflicts[i] = new long[scope[i].getDomain().length];
			}
		}

	}

	public ExtensionConstraint2D(final Variable[] scope, final Matrix2D matrix,
			final String name, final boolean shared) {
		super(scope, name);

		this.matrix = new MatrixManager2D(scope, matrix, shared);
		this.matrix.setTuple(tuple);
	}

	public boolean findValidTuple(final int variablePosition, final int index) {
		assert this.isInvolved(getVariable(variablePosition));

		return matrix.hasSupport(variablePosition, index);
	}

	public boolean revise(final int position, final int level) {
		final Variable variable = getVariable(position);

		assert !variable.isAssigned();

		boolean revised = false;

		for (int index = variable.getFirst(); index >= 0; index = variable
				.getNext(index)) {

			if (!findValidTuple(position, index)) {

				variable.remove(index, level);
				revised = true;
				setActive(true);

			}

		}

		return revised;
	}

	@Override
	public MatrixManager getMatrix() {
		return matrix;
	}

	@Override
	public int removeTuple(int[] tuple) {
		if (matrix.removeTuple(tuple)) {
			addConflict(tuple);
			return 1;
		}
		return 0;
	}

	@Override
	public boolean check() {
		return matrix.check();
	}

	public ExtensionConstraint2D deepCopy(final Collection<Variable> variables)
			throws CloneNotSupportedException {
		final ExtensionConstraint2D constraint = (ExtensionConstraint2D) super
				.deepCopy(variables);
		constraint.matrix = (MatrixManager2D) matrix.deepCopy(constraint
				.getScope(), constraint.tuple);
		return constraint;
	}

	@Override
	public int removeTuples(int[] fixed, int mobile, int[] values) {
		int removed = 0;
		for (int i : values) {
			fixed[mobile] = i;
			if (removeTuple(fixed) > 0
					&& getVariable(mobile).getRemovedLevel(i) > 0) {
				removed++;
			}
		}
		return removed;
	}

	@Override
	public boolean isSlow() {
		return false;
	}
}

File: src/main/scala/cspfj/constraint/extension/ExtensionConstraint2D.scala
/**
 * CSPFJ - CSP solving API for Java
 * Copyright (C) 2006 Julien VION
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

package cspfj.constraint.extension;

import cspfj.constraint.VariablePerVariable
import cspfj.Variable
import cspfj.constraint.TupleEnumerator
import scala.annotation.tailrec
import cspfj.Statistic
import cspfj.StatisticsManager

object ExtensionConstraint2D {
  @Statistic
  var checks = 0;
  @Statistic
  var presenceChecks = 0
}

final class ExtensionConstraint2D(
  scope: Array[Variable],
  private var matrix2d: Matrix2D,
  shared: Boolean)
  extends ConflictCount(scope, matrix2d, shared)
  with VariablePerVariable {

  private val GAIN_OVER_GENERAL = 3;

  private val MINIMUM_SIZE_FOR_LAST = 3 * java.lang.Long.SIZE;

  /**
   * No need for "last" data structure if domain sizes <=
   * MINIMUM_SIZE_FOR_LAST
   */
  private val residues =
    if (scope.map(_.dom.maxSize).max > MINIMUM_SIZE_FOR_LAST) {
      Array(new Array[Int](scope(0).dom.maxSize), new Array[Int](scope(1).dom.maxSize))
    } else null

  override def getEvaluation = (scope(0).dom.size * scope(1).dom.size) / GAIN_OVER_GENERAL

  override def simpleEvaluation = 2

  def reviseVariable(position: Int, mod: List[Int]) =
    !supportCondition(position) && scope(position).dom.filter(i => hasSupport(position, i))

  def removeTuple(tuple: Array[Int]) = {
    disEntail();
    set(tuple, false)
  }

  def removeTuples(base: Array[Int]) = tuples(base).count(removeTuple)

  override def toString = "ext2d(" + scope(0) + ", " + scope(1) + ")" + (if (isEntailed) " [entailed]" else "");

  def hasSupport(variablePosition: Int, index: Int) =
    if (residues == null) hasSupportNR(variablePosition, index);
    else hasSupportR(variablePosition, index);

  private def hasSupportR(variablePosition: Int, index: Int) = {
    controlResidue(variablePosition, index) || {
      val matrixBV = matrix2d.getBitVector(variablePosition, index);
      val intersection = scope(1 - variablePosition).dom.intersects(matrixBV)

      if (intersection >= 0) {
        ExtensionConstraint2D.checks += 1 + intersection;
        residues(variablePosition)(index) = intersection;
        true;
      } else {
        ExtensionConstraint2D.checks += matrixBV.realSize;
        false;
      }
    }
  }

  private def hasSupportNR(variablePosition: Int, index: Int) = {
    val matrixBV = matrix2d.getBitVector(variablePosition, index);
    val intersection = scope(1 - variablePosition).dom.intersects(matrixBV)

    if (intersection >= 0) {
      ExtensionConstraint2D.checks += 1 + intersection;
      true;
    } else {
      ExtensionConstraint2D.checks += matrixBV.realSize;
      false;
    }
  }

  private def controlResidue(position: Int, index: Int) = {
    val part = residues(position)(index)
    ExtensionConstraint2D.presenceChecks += 1
    (part != -1 && scope(1 - position).dom.intersects(
      matrix2d.getBitVector(position, index), part))
  }

  override def checkIndices(t: Array[Int]) = matrix.check(t)

  override def unshareMatrix() = {
    matrix2d = super.unshareMatrix().asInstanceOf[Matrix2D]
    matrix2d
  }

}
